adapters/beachfront/beachfront.go

```diff
@@ -111,177 +113,208 @@ type BeachfrontResponseSlot struct {
+
+	return r
+}
+
+func newBeachfrontVideoRequest() beachfrontVideoRequest {
+	r := beachfrontVideoRequest{}
+	r.Cur = append(r.Cur, "USD")
+	r.IsPrebid = true
+
+	return r
+}
+
+func getBeachfrontExtension(imp openrtb.Imp) (openrtb_ext.ExtImpBeachfront, error) {
+	var err error
+	var bidderExt adapters.ExtImpBidder
+	var beachfrontExt openrtb_ext.ExtImpBeachfront
+
+	if err = json.Unmarshal(imp.Ext, &bidderExt); err != nil {
+		return beachfrontExt, err
+	}
+
+	if err = json.Unmarshal(bidderExt.Bidder, &beachfrontExt); err != nil {
+		return beachfrontExt, err
 	}
 
-	return []*adapters.RequestData{{
-		Method:  "POST",
-		Uri:     uri,
-		Body:    reqJSON,
-		Headers: headers,
-	}}, errs
+	return beachfrontExt, err
 }
 
 /*
-We have received a prebid request. It needs to be converted to a beachfront request. This is complicated
-by the fact that we have different servers for video/display and they have different contracts.
+getBannerRequest, singular. A "Slot" is an "imp," and each Slot can have an AppId, so just one
+request to the beachfront banner endpoint gets all banner Imps.
 */
-func preprocess(req *openrtb.BidRequest, uri string) (BeachfrontRequests, []error, int) {
-	var beachfrontReqs BeachfrontRequests
-	var errs = make([]error, 0, len(req.Imp))
-	var imps int
+func getBannerRequest(request *openrtb.BidRequest) (beachfrontBannerRequest, []error) {
+	var beachfrontReq beachfrontBannerRequest
+	var errs = make([]error, 0, len(request.Imp))
 
-	if uri == BannerEndpoint {
-		beachfrontReqs.Banner, errs, imps = getBannerRequest(req)
-	} else {
-		// If there were any Video imps in the request, we have skipped to here.
-		beachfrontReqs.Video, errs, imps = getVideoRequest(req)
-	}
+	var slotIndex = 0
+	var impIndex = 0
 
-	return beachfrontReqs, errs, imps
-}
+	beachfrontReq = newBeachfrontBannerRequest()
 
-func getBannerRequest(req *openrtb.BidRequest) (BeachfrontBannerRequest, []error, int) {
-	var bannerImpsIndex = 0
-	var beachfrontReq = NewBeachfrontBannerRequest()
-	var errs = make([]error, 0, len(req.Imp))
-	var imps = 0
-
-	// step through the prebid request "imp" and inject into the beachfront request.
-	for _, imp := range req.Imp {
-		if imp.Banner != nil {
-			beachfrontReq.Slots = append(beachfrontReq.Slots, BeachfrontSlot{})
-			bannerImpsIndex = len(beachfrontReq.Slots) - 1
-
-			beachfrontReq.Slots[bannerImpsIndex].Sizes = append(beachfrontReq.Slots[bannerImpsIndex].Sizes, BeachfrontSize{})
-			for j := range imp.Banner.Format {
-				if j > 0 {
-					beachfrontReq.Slots[bannerImpsIndex].Sizes = append(beachfrontReq.Slots[bannerImpsIndex].Sizes, BeachfrontSize{})
-				}
-				beachfrontReq.Slots[bannerImpsIndex].Sizes[j].H = imp.Banner.Format[j].H
-				beachfrontReq.Slots[bannerImpsIndex].Sizes[j].W = imp.Banner.Format[j].W
-			}
+	for i := 0; i < len(request.Imp); i++ {
+		beachfrontReq.Slots = append(beachfrontReq.Slots, beachfrontSlot{})
+		slotIndex = len(beachfrontReq.Slots) - 1
```

Comment:
Is it possible for `slotIndex` to diverge from `i`?

Reply:
related - I noticed that impIndex was redundant in that loop, so eliminated it.

As for divergence, in preProcess, initial request in broken into two separate requests, one containing imps that are all guaranteed to contain a banner, the other all are guaranteed to contain a video. So the banner request is sent into getBannerRequest, it's count of imps is the the same as the count of slots that I should need in the beachfrontReq. If there is an error in getBeachfrontExtension, that slot will still in the stack, but it's a junk slot. I don't see a way that could happen, as the format of the extension would have to be wrong, and I don't see how that would get past validation, but saying it did, I would not want to include that slot, and at that point it _should_ diverge, but, no, it wasn't.

So I'm thinking that it should have the possibility of diverging in that case. I've got that where I think I want it. I'm not sure how to cause such an error in test cases, though. I'm able to manually throw an error 
`		if(i == 1) {
			err = errors.New("test error")
		}`

while running a big request on a test page with a debugging server build, I can see it diverge, and it seems to be doing so as expected - if I have 6 banner imps, and 2 are thrown a test error, I get 4 Slots in the final request.

I'm looking into how to test case this.

Or maybe I should forget about this error case that I don't think will ever happen and just index everything by "i" ?

Reply:
Short answer - it wasn't but it is now, in a controlled way, but I'm not sure it should.

Reply:
@hhhjort 
Do you think it is better as it is now, or better if I drop the whole diverging stack concept, drop slotIndex, make the Slots slice the same size as the Imp slice initially, and just index both by i?

Reply:
Which ever is more appropriate, I should also apply to getVideoRequests, as it also has what is currently two redundant indexes,  i and videoImpsIndex,  impCount that is never used, and tests if the Ext failed.

Reply:
I'm going to go with this being a good test to make, just in case. Also, if I'm doing it like that here, I should do similar, and eliminate some redundancies, in getVideoRequests.  So please don't merge this until I get that. Shouldn't take long at all.

If it is true that the params validation will _always_ catch errors in the Ext, and that the Ext unmarshaling will _never_ fail (???), then I could just drop the slotIndex, all the appending, etc, and just make the slices the same size initially and index them by "i". 

If you think that the later approach is better (Probably not as effected by increased size of Bids array), just let me know and I can get them both into that form instead.

Reply:
I audited getVideoRequests and LGTM